src/pop_manager.erl

```diff
@@ -0,0 +1,318 @@
+%% 
+%% Can run on its own in a loop, but is meant to be extended
+%% to have more structure, to serve as verifier or player.
+%% Those will be implemented in pop_verifier.erl and pop_player.erl
+%% 
+%% This has the code that responds to block requests.
+%% Current idea is that only verifiers should do that,
+%% though it is not inconcievable that players can also
+%% provide that information to each other.
+
+-module(pop_manager).
+
+-export([
+	 new/2,
+	 on_net_message/5,
+	 loop/1
+	]).
+
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("stdlib/include/assert.hrl").
+
+-include("potato_records.hrl").
+
+%% @doc Makes new container, using pop protocol config data,
+%% and external function hooks.
+%% 
+%% Current hooks are: 
+%% - send network message
+%% - on_new_block: react to new block added to chain
+
+new(PopConfigData, PopManagerConfig) ->
+    #pop_manager{
+       pop_chain = pop_chain:new(PopConfigData),
+       config = PopManagerConfig,
+       unbound_blocks = maps:new()
+      }.
+    
+
+%% Is the block known to us?
+%% 
+%% It is either in chain, pending as an unbound block,
+%% or unknown altogether.
+%% Returns in_chain, unbound, or unknown
+get_block_status(Hash, PopManager) ->
+    #pop_manager{
+       pop_chain = PC,
+       unbound_blocks = UB
+      } = PopManager,
+    
+
+    R1 = pop_chain:find_block_by_id(Hash, PC),
+    case R1 of
+	{ok, _} ->
+	    in_chain;
+	error ->
+	    R2 = maps:find(Hash, UB),
+	    case R2 of
+		{ok, _} ->
+		    unbound;
+		error ->
+		    unknown
+	    end
+    end.
+
+%% get previous block
+get_prev(Block, PC) ->
+    {ok, Prev} = pop_chain:find_block_by_id(maps:get(prev_id, Block), PC),
+    Prev.
+
+%% get n'th previous block, stop at genisys block
+get_nth_prev_block(N, Block, PC) ->
+    Ht = maps:get(height, Block),
+    if (Ht == 0) or (N == 0) ->
+	    Block;
+       (Ht > 0) and (N > 0) ->
+	    get_nth_prev_block(N-1, get_prev(Block, PC), PC)
+    end.
+
+setup_range_request(UnknownBlock, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    N = PopManager#pop_manager.config#pop_manager_config.request_range_backup,
+
+    KnownBlock1 = pop_chain:get_head_block(PC),
+    KnownBlock2 = get_nth_prev_block(N, KnownBlock1, PC),
+    
+    {maps:get(this_id, KnownBlock1), maps:get(this_id, KnownBlock2), maps:get(this_id, UnknownBlock)}.
+
+%% given unknown block request, and longest known block from another party
+%% make a best guess about which blocks are unknown to the other party
+%% and return a array of hashes of those blocks
+
+compute_block_hash_range(UnknownBlockHash, KnownBlockHash, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    UnknownBlock = pop_chain:find_block_by_id(UnknownBlockHash, PC),
+    KnownBlock = pop_chain:find_block_by_id(KnownBlockHash, PC),
+
+    compute_block_hash_range(UnknownBlock, KnownBlock, PC, []).
+
+compute_block_hash_range(UnknownBlock, KnownBlock, PC, AccList) ->
```

Comment:
what happens when unknown block doesn't have known block as a parent?

Reply:
this still works; this function basically finds the highest common ancestor